### PR TITLE
planner/cascades: introduce TransformationID in cascades planner

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 

This PR introduces `TransformationID` in cascades planner.  `TransformationID` is the handle of a `Transformation`. When we want to add a new Transformation rule, we should first add its ID, and create the rule in the `transformationRuleList` with the same order..

With the `TransformationID`, we can index the `Transforamtion` by its ID instead of its pointer address. 

### Check List 

Tests 

 - No code

Code changes

 - Has exported function/method change

### CODE REVIEW

File: planner/cascades/optimize.go [hunk 1]
```diff
@@ -31,21 +31,23 @@ var DefaultOptimizer = NewOptimizer()
 type Optimizer struct {
 	transformationRuleMap map[memo.Operand][]Transformation
 	implementationRuleMap map[memo.Operand][]ImplementationRule
+	patternMap            map[Transformation]*memo.Pattern
```
Comment (winoros) [inline planner/cascades/optimize.go]: What about adding a ID field to pattern and changing the map to `[]*memoPattern`?

File: planner/cascades/optimize.go [hunk 2]
```diff
@@ -66,6 +68,27 @@ func (opt *Optimizer) GetImplementationRules(node plannercore.LogicalPlan) []Imp
 	return opt.implementationRuleMap[memo.GetOperand(node)]
 }
 
+// GetPattern returns the Pattern of the given Transformation rule.
+// It returns the cached Pattern if possible. Otherwise, generate a new Pattern.
+func (opt *Optimizer) GetPattern(rule Transformation) *memo.Pattern {
+	if p, ok := opt.patternMap[rule]; ok {
+		return p
+	}
+	return rule.GetPattern()
```
Comment (lamxTyler) [inline planner/cascades/optimize.go]: Unable to reach here?

Comment (francis0407) [reply planner/cascades/optimize.go]: Yes, just in case.

**Pull request merged**